### PR TITLE
TypeScript compat changes for CRA and AF

### PR DESCRIPTION
**Pull request by flexdinesh**
PR includes changes for —
- Backwards compatibility changes for TS 3.7
- Compatibility with CRA and AF

### CODE REVIEW

File: src/lazy/index.tsx [hunk 1]
```diff
@@ -1,43 +1,20 @@
...
 import { PHASE } from '../constants';
 import { hash, displayNameFromId, isNodeEnvironment } from '../utils';
-import type { Asset, Manifest } from '../webpack';
+import { Asset, Manifest } from '../webpack';
 
 import { createComponentClient } from './components/client';
 import { createComponentServer } from './components/server';
 import { createDeferred } from './deferred';
 import { ClientLoader, Loader, ServerLoader } from './loader';
+import { LazyOptions, LazyComponent } from './types';
```
Comment (flexdinesh) [inline src/lazy/index.tsx:11]: Maybe I should rename this to `LazyComponent as LazyComponentType`

Comment (flexdinesh) [reply src/lazy/index.tsx:11]: Renamed the local component name instead

File: src/lazy/index.tsx [hunk 2]
```diff
@@ -1,48 +1,25 @@
...
 function lazyProxy<C extends ComponentType<any>>(
   loader: Loader<C>,
-  { defer = PHASE.PAINT, moduleId = '', ssr = true }: Options = {}
+  { defer = PHASE.PAINT, moduleId = '', ssr = true }: LazyOptions = {}
 ): LazyComponent<C> {
   const isServer = isNodeEnvironment();
   const dataLazyId = hash(moduleId);
 
-  const LazyComponent: FunctionComponent<ComponentProps<C>> = isServer
+  const LazyComponent_: FunctionComponent<ComponentProps<C>> = isServer
```
Comment (albertogasparin) [inline src/lazy/index.tsx:22]: I'm happy to rename this but really don't like the underscore ending. Can we do `LazyBase` maybe? Or `LazyInternal`?

**Review by albertogasparin — CHANGES_REQUESTED** [review]

**Comment by jackbrown** [comment]
> BTW looks like yesterday AF was upgraded to TS 3.9. Is this still needed then?

@albertogasparin IMO it is, as the previous behavior was hiding the issue rather than actually fixing it

**Comment by flexdinesh** [comment]
@jackbrown @albertogasparin CRA is still on TS 3.7. We need this until CRA is upgraded to TS 3.8 or later.

File: src/lazy/index.tsx (hunk 2 above)
Comment (flexdinesh) [reply src/lazy/index.tsx:22]: Ah yeah I quickly push in a rename.

Comment (flexdinesh) [reply src/lazy/index.tsx]: Changes pushed. `LazyComponent_` is now `LazyInternal`.

**Review by albertogasparin — APPROVED** [review]

File: src/index.ts [hunk 3]
```diff
@@ -11,15 +10,11 @@ export {
   LoaderError,
 } from './lazy';
 
-export type {
-  ClientLoader,
-  Loader,
-  Options as LazyOptions,
-  ServerLoader,
-} from './lazy';
+export { ClientLoader, Loader, LazyOptions, ServerLoader } from './lazy';
```
Comment (MonicaOlejniczak) [inline src/index.ts:13]: This should also export `LazyComponent`, I must've missed it in initial implementations since Jira is using flow types

**Review by MonicaOlejniczak — CHANGES_REQUESTED** [review]
It does seem like CRA changed recently so that the types exported in the distribution no longer result in warnings, and then errors. However, I believe this will still result in build warnings in Jira, which is not desirable. Has there been any other investigations for achieving the same result without the warnings?

File: src/index.ts (hunk 3 above)
Comment (flexdinesh) [reply src/index.ts]: @MonicaOlejniczak I have exported the `LazyComponent` as well from index file and force pushed. Thanks for catching that. I experimented with CRA latest version and it doesn't seem like CRA handles `export type` syntax without errors 'cos it's on TS v3.7.2.

Comment (flexdinesh) [reply src/index.ts]: What type of warnings do you run into in Jira?

Comment (MonicaOlejniczak) [reply src/index.ts]: When you don't use the `type` keyword, the build output contains references to type exports in the library files (e.g. `dist/esm/index.js`). These type exports cannot be resolved as they are stripped out during the build. Here's an example:

```
// esm/index.js
export { ClientLoader, Loader, LazyOptions, ServerLoader } from './lazy';

// esm/lazy/index.js
export { Asset, Manifest, LazyOptions, LazyComponent };

// esm/lazy/types.js
// empty file -- LazyComponent and LazyOptions do not exist

```

**Review by MonicaOlejniczak — APPROVED** [review]
LGTM

**Pull request merged**